File: FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/node/gson/MsgBlockDeserializer.java
/*
 * © 2016-2019 FlowCrypt Limited. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.api.retrofit.node.gson;

import com.flowcrypt.email.api.retrofit.response.model.node.BaseMsgBlock;
import com.flowcrypt.email.api.retrofit.response.model.node.DecryptErrorMsgBlock;
import com.flowcrypt.email.api.retrofit.response.model.node.MsgBlock;
import com.flowcrypt.email.api.retrofit.response.model.node.PublicKeyMsgBlock;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;

/**
 * This realization helps parse a right variant of {@link MsgBlock}
 *
 * @author Denis Bondarenko
 * Date: 3/26/19
 * Time: 9:36 AM
 * E-mail: DenBond7@gmail.com
 */
public class MsgBlockDeserializer implements JsonDeserializer<MsgBlock> {
  @Override
  public MsgBlock deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
      throws JsonParseException {
    JsonObject jsonObject = json.getAsJsonObject();

    MsgBlock.Type type = context.deserialize(jsonObject.get(MsgBlock.TAG_TYPE), MsgBlock.Type.class);

    switch (type) {
      case PUBLIC_KEY:
        return context.deserialize(json, PublicKeyMsgBlock.class);

      case DECRYPT_ERROR:
        return context.deserialize(json, DecryptErrorMsgBlock.class);

      default:
        return context.deserialize(json, BaseMsgBlock.class);
    }
  }
}

File: FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/node/gson/MsgBlockDeserializer.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.api.retrofit.node.gson

import com.flowcrypt.email.api.retrofit.response.model.node.DecryptErrorMsgBlock
import com.flowcrypt.email.api.retrofit.response.model.node.DecryptedAttMsgBlock
import com.flowcrypt.email.api.retrofit.response.model.node.GenericMsgBlock
import com.flowcrypt.email.api.retrofit.response.model.node.MsgBlock
import com.flowcrypt.email.api.retrofit.response.model.node.PublicKeyMsgBlock
import com.flowcrypt.email.jetpack.viewmodel.PrivateKeysViewModel
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import java.lang.reflect.Type

/**
 * This realization helps parse a right variant of [MsgBlock]
 *
 * @author Denis Bondarenko
 * Date: 3/26/19
 * Time: 9:36 AM
 * E-mail: DenBond7@gmail.com
 */
class MsgBlockDeserializer : JsonDeserializer<MsgBlock> {
  override fun deserialize(
    json: JsonElement,
    typeOfT: Type,
    context: JsonDeserializationContext
  ): MsgBlock? {
    val jsonObject = json.asJsonObject

    val msgBlock: MsgBlock? = when (
      context.deserialize<MsgBlock.Type>(
        jsonObject.get(MsgBlock.TAG_TYPE),
        MsgBlock.Type::class.java
      )
    ) {
      MsgBlock.Type.PUBLIC_KEY -> context.deserialize(json, PublicKeyMsgBlock::class.java)
      MsgBlock.Type.DECRYPT_ERROR -> context.deserialize(json, DecryptErrorMsgBlock::class.java)
      MsgBlock.Type.DECRYPTED_ATT -> context.deserialize(json, DecryptedAttMsgBlock::class.java)
      else -> context.deserialize(json, GenericMsgBlock::class.java)
    }

    if (msgBlock?.type == MsgBlock.Type.PUBLIC_KEY) {
      // check https://github.com/FlowCrypt/flowcrypt-android/issues/911 to see details
      if (msgBlock.content?.length ?: 0 > PrivateKeysViewModel.MAX_SIZE_IN_BYTES) {
        return null
      }
    }

    return msgBlock
  }
}
